flake.nix

```diff
@@ -0,0 +1,115 @@
+              "-L${avrlibc}/avr/lib/avr35"
+              "-B${avrlibc}/avr/lib/avr51"
+              "-L${avrlibc}/avr/lib/avr51"
+            ];
+
+            # Builds the python env based on nix/pyproject.toml and
+            # nix/poetry.lock Use the "poetry update --lock", "poetry add
+            # --lock" etc. in the nix folder to adjust the contents of those
+            # files if the requirements*.txt files change
+            pythonEnv = poetry2nix.mkPoetryEnv {
+              projectDir = ./util/nix;
+              overrides = poetry2nix.overrides.withDefaults (self: super: {
+                pillow = super.pillow.overridePythonAttrs (old: {
+                  # Use preConfigure from nixpkgs to fix library detection issues and
+                  # impurities which can break the build process; this also requires
+                  # adding propagatedBuildInputs and buildInputs from the same source.
+                  propagatedBuildInputs = (old.buildInputs or [ ]) ++ pkgs.python3.pkgs.pillow.propagatedBuildInputs;
+                  buildInputs = (old.buildInputs or [ ]) ++ pkgs.python3.pkgs.pillow.buildInputs;
+                  preConfigure = (old.preConfigure or "") + pkgs.python3.pkgs.pillow.preConfigure;
+                });
+                qmk = super.qmk.overridePythonAttrs (old: {
+                  # Allow QMK CLI to run "qmk" as a subprocess (the wrapper changes
+                  # $PATH and breaks these invocations).
+                  dontWrapPythonPrograms = true;
+                });
+              });
+            };
+          in
+          pkgs.mkShell {
+            name = "qmk-firmware";
+
+            buildInputs = with pkgs;[
+              clang-tools
+              dfu-programmer
+              dfu-util
+              diffutils
+              git
+              pythonEnv
+              niv
+            ]
+            ++ lib.optional avr [
+              pkgsCross.avr.buildPackages.binutils
+              pkgsCross.avr.buildPackages.gcc8
+              avrlibc
+              avrdude
+            ]
+            ++ lib.optional arm [ gcc-arm-embedded ]
+            ++ lib.optional teensy [ teensy-loader-cli ];
+
+            AVR_CFLAGS = lib.optional avr avr_incflags;
+            AVR_ASFLAGS = lib.optional avr avr_incflags;
+            shellHook = ''
+              # Prevent the avr-gcc wrapper from picking up host GCC flags
+              # like -iframework, which is problematic on Darwin
+              unset NIX_CFLAGS_COMPILE_FOR_TARGET
+            '';
+
+            # pass through no* attributes so they
+            # can be combined (in any order) like this:
+            # nix develop .#noAvr.noArm.noTeensy
+            passthru =
+              {
+                noAvr = (makeShell { inherit arm teensy; avr = false; });
+                noArm = (makeShell { inherit avr teensy; arm = false; });
+                noTeensy = (makeShell { inherit avr arm; teensy = false; });
+              };
```

Comment:
Interesting trick; so officially `devShells` should be single-level, but apparently having subpackages this way works.